src/Scout/ScoutEngine.php

```diff
@@ -0,0 +1,542 @@
+        assert($model instanceof Model, new TypeError(sprintf('Argument #3 ($model) must be of type %s, %s given', Model::class, get_debug_type($model))));
+
+        if (! $results) {
+            return LazyCollection::make($model->newCollection());
+        }
+
+        $objectIds = array_column($results, '_id');
+        $objectIdPositions = array_flip($objectIds);
+
+        return $model->queryScoutModelsByIds(
+            $builder,
+            $objectIds,
+        )->cursor()->filter(function ($model) use ($objectIds) {
+            return in_array($model->getScoutKey(), $objectIds);
+        })->map(function ($model) use ($results, $objectIdPositions) {
+            $result = $results[$objectIdPositions[$model->getScoutKey()]] ?? [];
+
+            foreach ($result as $key => $value) {
+                if (substr($key, 0, 1) === '_') {
+                    $model->withScoutMetadata($key, $value);
+                }
+            }
+
+            return $model;
+        })->sortBy(function ($model) use ($objectIdPositions) {
+            return $objectIdPositions[$model->getScoutKey()];
+        })->values();
+    }
+
+    /**
+     * Create the MongoDB Atlas Search index.
+     *
+     * Accepted options:
+     *  - wait: bool, default true. Wait for the index to be created.
+     *
+     * @see Engine::createIndex()
+     *
+     * @param string            $name    Collection name
+     * @param array{wait?:bool} $options
+     */
+    #[Override]
+    public function createIndex($name, array $options = []): void
+    {
+        assert(is_string($name), new TypeError(sprintf('Argument #1 ($name) must be of type string, %s given', get_debug_type($name))));
+
+        // Ensure the collection exists before creating the search index
+        $this->database->createCollection($name);
+
+        $collection = $this->database->selectCollection($name);
+        $collection->createSearchIndex(
+            self::DEFAULT_DEFINITION,
+            ['name' => self::INDEX_NAME],
+        );
+
+        if ($options['wait'] ?? true) {
+            $this->wait(function () use ($collection) {
+                $indexes = $collection->listSearchIndexes([
+                    'name' => self::INDEX_NAME,
+                    'typeMap' => ['root' => 'bson'],
+                ]);
+
+                return $indexes->current() && $indexes->current()->status === 'READY';
+            });
+        }
+    }
+
+    /**
+     * Delete a "search index", i.e. a MongoDB collection.
+     *
+     * @see Engine::deleteIndex()
+     */
+    #[Override]
+    public function deleteIndex($name): void
+    {
+        assert(is_string($name), new TypeError(sprintf('Argument #1 ($name) must be of type string, %s given', get_debug_type($name))));
+
+        $this->database->selectCollection($name)->drop();
```

Comment:
Alternatively use [`Database::dropCollection()`](https://www.mongodb.com/docs/php-library/current/reference/method/MongoDBDatabase-dropCollection/):

```suggestion
        $this->database->dropCollection($name);
```